web/packages/core/src/ruffle-player.ts

```diff
@@ -821,6 +828,39 @@ export class RufflePlayer extends HTMLElement {
         }
     }
 
+    private virtualKeyboardInput() {
+        const input = this.virtualKeyboard;
+        const initialValue = input.defaultValue;
+        const initialValueLength = initialValue.length;
+        const isBackspace = input.value.length <= initialValueLength;
+        const chars = isBackspace
+            ? ["Backspace"]
+            : [...input.value.slice(initialValueLength)];
+
+        for (const char of chars) {
+            for (const eventType of ["keydown", "keyup"]) {
+                this.dispatchEvent(
+                    new KeyboardEvent(eventType, {
+                        key: char,
+                        bubbles: true,
+                    })
+                );
+            }
+        }
+        input.value = initialValue;
+    }
+    private openVirtualKeyboard(): void {
+        if (!("listener" in this.virtualKeyboard.dataset)) {
```

Comment:
My feeling is that adding an event listener unconditionally isn't much costly, but it would simplify things a lot.

Reply:
Would it not trigger all of them though? So opening the virtual keyboard with the context menu twice types each character twice?

Reply:
And simplifying things "a lot" seems like a bit of an exaggeration, as wouldn't it just remove the conditional that checks for the data-* element and the line that sets the data-* element?

Reply:
Just tested an unconditional statement here. It's fairly unpredictable what exactly it does because of race conditions. But usually if you open the virtual keyboard multiple times, it seem like typing will stop doing anything or type one or more backspaces. The input event needs to only fire once.

Reply:
Or wait, you mean add the unconditional event listener to https://github.com/ruffle-rs/ruffle/pull/9461/files#diff-467a580c71d4e3b96032b54a2a292d485f3b83f87ba2478daf693bac6d532965R488

Reply:
Oh yeah, sorry if I wasn't clear. I meant adding the event listener during the one-time initialization.